src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/MsaPassthroughHandler.cs

```diff
@@ -0,0 +1,146 @@
+            _logger.Info("WAM MSA-PT: PassThrough is enabled? " + enabled);
+
+            return enabled;
+        }
+
+        public async Task<string> FetchTransferTokenAsync(AuthenticationRequestParameters authenticationRequestParameters, WebAccountProvider accountProvider)
+        {
+            _logger.Verbose("WAM MSA-PT - fetching transfer token");
+            string transferToken = await FetchMsaPassthroughTransferTokenAsync(
+                authenticationRequestParameters, accountProvider)
+                .ConfigureAwait(false);
+
+            return transferToken;
+        }
+
+        public void AddTransferTokenToRequest(
+            Windows.Security.Authentication.Web.Core.WebTokenRequest webTokenRequest,
+            string transferToken)
+        {
+            if (!string.IsNullOrEmpty(transferToken))
+            {
+                webTokenRequest.Properties.Add("SamlAssertion", transferToken);
+                webTokenRequest.Properties.Add("SamlAssertionType", "SAMLV1");
+            }
+        }
+
+        private async Task<string> FetchMsaPassthroughTransferTokenAsync(
+           AuthenticationRequestParameters authenticationRequestParameters,
+           WebAccountProvider accountProvider)
+        {
+            // step 1 - get a response from the MSA provider, just to have a WebAccount
+            _logger.Info("WAM MSA-PT: Making initial call to MSA provider");
+            WebAccount msaPtWebAccount = await FetchWebAccountFromMsaAsync(
+                authenticationRequestParameters,
+                accountProvider).ConfigureAwait(false);
+
+            // step 2 - get a trasnfer token 
+            _logger.Info("WAM MSA-PT: Getting transfer token");
+            string transferToken = await FetchTransferTokenAsync(
+                accountProvider,
+                msaPtWebAccount,
+                authenticationRequestParameters.ClientId).ConfigureAwait(true);
+
+            return transferToken;
+        }
+
+        private async Task<WebAccount> FetchWebAccountFromMsaAsync(
+            AuthenticationRequestParameters authenticationRequestParameters, WebAccountProvider accountProvider)
+        {
+            // This response has an v1 MSA Access Token, which MSAL should expose to the user
+            var webTokenRequestMsa = await _msaPlugin.CreateWebTokenRequestAsync(
+                     accountProvider,
+                     authenticationRequestParameters,
+                     isForceLoginPrompt: false,
+                     isInteractive: true,
+                     isAccountInWam: false)
+                    .ConfigureAwait(false);
+
+            WamAdapters.AddMsalParamsToRequest(authenticationRequestParameters, webTokenRequestMsa);
+
+            var webTokenResponseMsa = await _wamProxy.RequestTokenForWindowAsync(_parentHandle, webTokenRequestMsa)
+                .ConfigureAwait(true);
+
+            if (!webTokenResponseMsa.ResponseStatus.IsSuccessStatus())
```

Comment:
>webTokenResponseMsa [](start = 17, length = 19)

can this be null?

Reply:
webTokenResponseMsa cannot be null - it's a bit like our MsalTokenResponse - it conflates error and success state. In this case, they have an enum to tell the actual status.